openedx/core/djangoapps/schedules/tasks.py

```diff
@@ -223,53 +236,156 @@ def make_message_type(self, day_offset):
 ScheduleCourseUpdate = ScheduleCourseUpdate.task_instance
 
 
-class ScheduleCourseNextSectionUpdate(ScheduleMessageBaseTask):  # lint-amnesty, pylint: disable=missing-class-docstring
-    enqueue_config_var = 'enqueue_course_update'
-    log_prefix = COURSE_NEXT_SECTION_UPDATE_LOG_PREFIX
-    resolver = resolvers.CourseNextSectionUpdate
-    async_send_task = _course_update_schedule_send
+class SelfPacedCourseMessageBaseTask(ScheduleMessageBaseTask):
+    """
+    Base class for Self paced course schedule tasks that create subtasks.
+    """
+    enqueue_config_var = ''
+    log_prefix = ''
+    resolver = None  # define in subclass
+    async_send_task = None
     task_instance = None
 
     @classmethod
-    def enqueue(cls, site, current_date, day_offset, override_recipient_email=None):  # lint-amnesty, pylint: disable=missing-function-docstring
+    def should_process_course(cls, course_key):  # pylint: disable=unused-argument
+        """
+        Check if course should be processed for scheduled messages.
+        Override in child class if this check is required.
+        """
+        return True
+
+    @classmethod
+    def calculate_target_datetime(cls, effective_date, day_offset):  # pylint: disable=unused-argument
+        """
+        Get target datetime for given effective_date and day_offset.
+        """
+        raise NotImplementedError
+
+    def make_message_type(self):
+        """ Make schedule message type. """
+        raise NotImplementedError
+
+    @classmethod
+    def get_task_args(cls, site, effective_date, target_datetime, course_key, override_recipient_email):
+        """ Get async task args. """
+        return (
+            site.id,
+            serialize(effective_date),
+            serialize(target_datetime),
+            str(course_key),  # Needs to be a string for celery to properly process
+            override_recipient_email,
+        )
```

Comment:
If I understand correctly, we introduced this to pass `target_datetime` as the `effective_datetime` `ScheduleCourseNextSectionUpdate`.

Could we remove this and change `calculate_target_datetime` to something like `calculate_dates`, which would return `effective_date, target_date`? It will be easier to understand this way.